rfc/monorepo.md

```diff
@@ -0,0 +1,557 @@
+
+### Subtree Merges and Contributions From Existing Forks
+
+This section details how to integrate changes between a single project repository (e.g. an existing OSS project or a private fork) and a subtree-ified monorepo of all Spinnaker projects.
+
+Heavy use is made of Git's [`subtree` merge strategy](https://git-scm.com/docs/merge-strategies#Documentation/merge-strategies.txt-subtree).
+
+Expects that you have [SSH-based Github authentication](https://docs.github.com/en/authentication/connecting-to-github-with-ssh) configured on your local machine.
+
+#### Creation of the OSS Monorepo
+
+A subtree-ified monorepo of the entire Spinnaker project can be created with the following script, following the `git init` command in your folder of choice:
+
+```sh
+#!/usr/bin/env sh
+
+# Subtree clones
+git subtree add -P clouddriver git@github.com:spinnaker/clouddriver.git master
+git subtree add -P deck git@github.com:spinnaker/deck.git master
+git subtree add -P deck-kayenta git@github.com:spinnaker/deck-kayenta.git master
+git subtree add -P echo git@github.com:spinnaker/echo.git master
+git subtree add -P fiat git@github.com:spinnaker/fiat.git master
+git subtree add -P front50 git@github.com:spinnaker/front50.git master
+git subtree add -P gate git@github.com:spinnaker/gate.git master
+git subtree add -P halyard git@github.com:spinnaker/halyard.git master
+git subtree add -P igor git@github.com:spinnaker/igor.git master
+git subtree add -P kayenta git@github.com:spinnaker/kayenta.git master
+git subtree add -P keel git@github.com:spinnaker/keel.git master
+git subtree add -P kork git@github.com:spinnaker/kork.git master
+git subtree add -P orca git@github.com:spinnaker/orca.git master
+git subtree add -P rosco git@github.com:spinnaker/rosco.git master
+
+git fetch --all
+```
+
+Congratulations!  All Git setup for the new monorepo is done!
+
+#### Transitional Merges to the OSS Monorepo
+
+Work will continue on the individual projects for some time.  Fortunately, the new monorepo can be easily updated from the individual projects with the following process:
+
+1. Add remotes
+    - `git remote add <individual_project_name> <git_url>`
+    - Repeat for all projects
+
+1. Merge changes from the individual repo to the monorepo using the `subtree` strategy
+    - `git merge -X subtree=<subtree> <remote_name>/<branch>`
+    - Repeat for each remote project and each branch that needs reintegration
+
+This is a simple shell script that implements a process like the above:
+
+<details>
+<summary>Expand pull.sh</summary>
+
```

Comment:
Wondering if this should be a git check on merge kinda thing to make sure things are updated?  Just ... would hate to have it pure manual kinda thing.  Maybe a gradle task sorta like the spotless rules to verify subtrees are there/updated/etc.

Reply:
We could, but it's a purely transitional process - once active development has moved to the monorepo, it won't be needed anymore.  

It's also possible that manual conflict resolution could be required if there is new development in the monorepo not present in the individual projects.

My suggestion would probably be a GHA manual or scheduled trigger that runs and auto-commits successful merges, with the possibility of a human running it also available.